### PR TITLE
Deny service ClusterIP update from `None`

### LINKED ISSUE #33029: Service can change from headless -> non-headless and never gets an IP allocated
This is probably not that hard to fix, but it will want an e2e case to cover it.

```
$ cat /tmp/hdls.yaml 
apiVersion: v1
kind: Service
metadata:
  name: headless
spec:
  clusterIP: None
  ports:
  - port: 80
  selector:
    run: headless

$ k create -f /tmp/hdls.yaml 
service "headless" created

$ k get svc headless -o yaml | grep clusterIP
  clusterIP: None

$ k get svc headless -o yaml | sed '/clusterIP/d' | k replace -f -
service "headless" replaced

$ k get svc headless -o yaml | grep clusterIP

$ sleep 10; k get svc headless -o yaml | grep clusterIP

$
```

Additionally, if I manually assign an IP it is not verified:

```
$ k get svc
NAME         CLUSTER-IP   EXTERNAL-IP     PORT(S)   AGE
headless     10.0.0.93    <none>          80/TCP    4m
hostnames    10.0.0.93    104.197.1.173   80/TCP    4d
kubernetes   10.0.0.1     <none>          443/TCP   6d
```

And kube-proxy is VERY confused.

### PR DESCRIPTION
**Pull request by nebril**
**What this PR does / why we need it**: Headless service should not be transformed into a service with ClusterIP, therefore update of this field if it's set to `None` is disallowed.

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #33029

**Release note**:

``` release-note
Changing service ClusterIP from `None` is not allowed anymore.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34094)

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2575,6 +2575,10 @@ func ValidateServiceUpdate(service, oldService *api.Service) field.ErrorList {
 		allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
 	}
 
+	if oldService.Spec.ClusterIP == api.ClusterIPNone {
+		allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
```
Comment (thockin) [inline pkg/api/validation/validation.go]: This block is identical to above.  Can we just flatten this to a unilateral `ValidateImmutableField` call?  I can't think of any valid update of this field.

CREATE sets "" -> allocated
UPDATE A -> B is disallowed
UPDATE A -> "" is disallowed
UPDATE A -> "None" is disallowed
UPDATE None -> B is disallowed
UPDATE None -> "" is disallowed
UPDATE None -> "None" is disallowed

We should never every see it as ""

Comment (nebril) [inline pkg/api/validation/validation.go]: Makes sense, I will update it.

Comment (nebril) [inline pkg/api/validation/validation.go]: @thockin what about the case in https://github.com/kubernetes/kubernetes/blob/caad4db78be6ed/test/e2e/dns.go#L481 ? It seems like you can change ClusterIP when changing type of service. I think my check should prevail.

Comment (thockin) [inline pkg/api/validation/validation.go]: oh, good point.

OK, can we then make this code more obvious and comment it heavily to explain which transitions are being caught and which are not?

Comment (nebril) [inline pkg/api/validation/validation.go]: Sure thing.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -2571,7 +2571,9 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
 func ValidateServiceUpdate(service, oldService *api.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
 
-	if api.IsServiceIPSet(oldService) {
+	// ClusterIP should be mutable only for empty value, to allow assigning ClusterIP
+	// for service changing type to one that requires ClusterIP
+	if oldService.Spec.ClusterIP != "" {
```
Comment (thockin) [inline pkg/api/validation/validation.go]: What about the other direction?  We've been a little lax in the definition here.  Do we think it is valid to change from ClusterIP to ExternalName types?

Comment (thockin) [inline pkg/api/validation/validation.go]: I wonder if it makes more sense to refactor this into a top-level switch on service.Spec.Type and just handle each case in a more obvious way...  Thoughts?

Comment (nebril) [inline pkg/api/validation/validation.go]: Seems like it. Will try to carefully think of all cases.

**Comment by thockin** [comment]
Let me know when ready for review

**Comment by nebril** [comment]
@thockin I think it's ready, I covered most of the cases in tests, everything seems to work fine.

**Comment by nebril** [comment]
@thockin rebased, pinging for review

**Comment by nebril** [comment]
Agreed, IP should be allocated on such change. Will see if there is an issue for that, if not I will create one.

**Comment by nebril** [comment]
OK, I found #35354

**Pull request merged**